- Exercici 1

Crea almenys dos models de regressió diferents per intentar predir el millor possible el preu de les vivendes (MEDV) de l'arxiu adjunt.

- Exercici 2

Compara’ls en base al MSE i al R2.

- Exercici 3

Entrena’ls utilitzant els diferents paràmetres que admeten per intentar millorar-ne la predicció.

- Exercici 4

Compara el seu rendiment emprant l’aproximació traint/test o emprant totes les dades (validació interna).

- Exercici 5

No facis servir la variable del nombre d'habitacions (RM) a l’hora de fer prediccions.


***Sobre cada model generat es realitzen cada un dels exercicis sol·licitats***

In [101]:
import pandas as pd

pd.read_csv('housing_data.csv')

Columnes = ['CRIM','ZN','INDUS','CHAS','NOX','RM','AGE','DIS','RAD','TAX','PTRATIO','B','LSTAT','MEDV']

Boston_df = pd.read_csv('housing_data.csv', names = Columnes)

Boston_df

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273.0   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273.0   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273.0   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273.0   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273.0   

     PTRATIO       B  LSTAT  MEDV  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
4       18.7  396.90   5.33  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99   9.67  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  

[506 rows x 14 columns]

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

### Model 1: Regressió lineal amb 'Statsmodels'


In [102]:
# Definim les variables independents (feature_names) i la variable depenent (objectiu / target)

variables = Boston_df.drop('MEDV', axis = 1)
objectiu = Boston_df['MEDV']

print(variables)
print(objectiu)

        CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273.0   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273.0   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273.0   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273.0   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273.0   

     PTRATIO       B  LSTAT  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       1

In [103]:
import statsmodels.api as sm

# Ajustem el model amb mes d'una variable

x = variables[['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO','B', 'LSTAT']]
y = objectiu   # ['MEDV']

# Calculem

model = sm.OLS(y, x).fit() 
prediccions = model.predict(x)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                   MEDV   R-squared (uncentered):                   0.959
Model:                            OLS   Adj. R-squared (uncentered):              0.958
Method:                 Least Squares   F-statistic:                              891.3
Date:                Thu, 23 Mar 2023   Prob (F-statistic):                        0.00
Time:                        00:46:03   Log-Likelihood:                         -1523.8
No. Observations:                 506   AIC:                                      3074.
Df Residuals:                     493   BIC:                                      3128.
Df Model:                          13                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
CRIM          -0.0929      0.034     -2.699      0.007      -0.161      -0.025
ZN             0.0487      0.014      3.382      0.001       0.020       0.077
INDUS         -0.0041      0.064     -0.063      0.950      -0.131       0.123
CHAS           2.8540      0.904      3.157      0.002       1.078       4.630
NOX           -2.8684      3.359     -0.854      0.394      -9.468       3.731
RM             5.9281      0.309     19.178      0.000       5.321       6.535
AGE           -0.0073      0.014     -0.526      0.599      -0.034       0.020
DIS           -0.9685      0.196     -4.951      0.000      -1.353      -0.584
RAD            0.1712      0.067      2.564      0.011       0.040       0.302
TAX           -0.0094      0.004     -2.395      0.017      -0.017      -0.002
PTRATIO       -0.3922      0.110     -3.570      0.000      -0.608      -0.176
B              0.0149      0.003      5.528      0.000       0.010       0.020
LSTAT         -0.4163      0.051     -8.197      0.000      -0.516      -0.317
==============================================================================
Omnibus:                      204.082   Durbin-Watson:                   0.999
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1374.225
Skew:                           1.609   Prob(JB):                    3.90e-299
Kurtosis:                      10.404   Cond. No.                     8.50e+03
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[3] The condition number is large, 8.5e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [104]:
# Calculem el MSE:

from sklearn.metrics import mean_squared_error
import statsmodels.api as sm

mse = mean_squared_error(y, prediccions)

print("MSE:", mse)

MSE: 24.166099330126496


### Comentaris:

- El R2 = 0.959 ens indica que el model presenta un bon ajust, que les variables independents están altament relacionades amb la variable depenent.
- Probablement obtenim un valor R2 tant elevat perque hem considerat totes les variables independents (13) per generar el model.
- En el cas que consideresim variable amb un coeficient de correlació baix com `CRIM' o 'ZN' el valor de R2 sería molt més baix, significant que aquestes variables no están altament relacionades amb la variable depenent.
- Observant els ceficients de correlació (coef) podriem intuir quines variables podrien estar més relacionades amb 'MEDV'.
- Podriem utilitzar les variables 'RM', 'LSTAT', 'NOX' i/o 'DIS' per predir 'MEDV'

---------------------------------------------------------------------------------------------------------------------------

In [105]:
# Entrenem el model amb regularització LASSO

  # LASSO: especielment utilitzat quan s'espera que, de totes les varibales independents, només siguin rellevants unes poques.   

from sklearn.linear_model import Lasso

model_lasso_sm = Lasso(alpha=1.0)

model_lasso_sm.fit(x, y)

Lasso()

In [106]:
# Prediccions (LASSO):

Prediccions_sm_Lasso = model_lasso_sm.predict(x)

print(Prediccions_sm_Lasso[:5]) # mostrem els primers 5 valors obtinguts

[30.99753918 25.77681736 29.98601449 29.51799813 28.03248999]


In [107]:
# Calculem el R2 (LASSO)

model_lasso_sm.score(x,y)

0.6825842212709925

### Comentaris:

- Al limitar la magnitud dels coeficients del model de regressió pot reduïr l'ajust exceciu del model
- En aquest sentit, pot reduir-se la capacitat del model per ajustar-se perfectament a les dades, el que pot comportar una reducció del R2, tal com observem en l'ajust amb LASSO.
- No obstant, el fet de tenir una R2 menor no implica necessariament que el model sigui pitjor, sino que s'han de tenir en compte altres aspectes.

----------------------------------------------------------------------------------------------------------------------------

In [108]:
# Fem la partició del conjunt de dades en 'Train' i 'Test'

from sklearn.model_selection import train_test_split

train_x, test_x, train_y, test_y = train_test_split(x, y, test_size = 0.20,random_state = 42)

print('Característiques Train_x:', train_x.shape)
print('Característiques Train_y:', train_y.shape)
print('Característiques Test_x:', test_x.shape)
print('Característiques Test_y:', test_y.shape)

# En 'Train' tenim el 80% dels valors (404 files) i en 'Test' el 20% (102 files)


Característiques Train_x: (404, 13)
Característiques Train_y: (404,)
Característiques Test_x: (102, 13)
Característiques Test_y: (102,)


In [109]:
import statsmodels.api as sm

# Ajustem el model 'Train' amb mes d'una variable

Ajust_train_x = train_x [['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO','B', 'LSTAT']]
Ajust_train_y = train_y   # ['MEDV']

# Calculem model 'Train'

model = sm.OLS(Ajust_train_y, Ajust_train_x).fit() 
prediccions = model.predict(Ajust_train_x)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                   MEDV   R-squared (uncentered):                   0.962
Model:                            OLS   Adj. R-squared (uncentered):              0.960
Method:                 Least Squares   F-statistic:                              755.7
Date:                Thu, 23 Mar 2023   Prob (F-statistic):                   1.13e-267
Time:                        00:46:05   Log-Likelihood:                         -1208.5
No. Observations:                 404   AIC:                                      2443.
Df Residuals:                     391   BIC:                                      2495.
Df Model:                          13                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
CRIM          -0.1005      0.036     -2.799      0.005      -0.171      -0.030
ZN             0.0342      0.017      2.068      0.039       0.002       0.067
INDUS          0.0336      0.070      0.482      0.630      -0.103       0.171
CHAS           2.7397      0.989      2.771      0.006       0.796       4.683
NOX           -5.1551      3.706     -1.391      0.165     -12.440       2.130
RM             6.1960      0.338     18.352      0.000       5.532       6.860
AGE           -0.0111      0.015     -0.737      0.462      -0.041       0.018
DIS           -1.0200      0.218     -4.669      0.000      -1.450      -0.591
RAD            0.1388      0.075      1.855      0.064      -0.008       0.286
TAX           -0.0078      0.004     -1.790      0.074      -0.016       0.001
PTRATIO       -0.4541      0.119     -3.815      0.000      -0.688      -0.220
B              0.0168      0.003      5.728      0.000       0.011       0.023
LSTAT         -0.4259      0.055     -7.699      0.000      -0.535      -0.317
==============================================================================
Omnibus:                      154.481   Durbin-Watson:                   2.089
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1082.869
Skew:                           1.452   Prob(JB):                    7.21e-236
Kurtosis:                      10.477   Cond. No.                     8.46e+03
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[3] The condition number is large, 8.46e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [110]:
import statsmodels.api as sm

# Ajustem el model 'Test' amb mes d'una variable

Ajust_test_x = test_x [['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO','B', 'LSTAT']]
Ajust_test_y = test_y   # ['MEDV']

# Calculem model 'Test'

model = sm.OLS(Ajust_test_y, Ajust_test_x).fit() 
prediccions = model.predict(Ajust_test_x)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                   MEDV   R-squared (uncentered):                   0.955
Model:                            OLS   Adj. R-squared (uncentered):              0.949
Method:                 Least Squares   F-statistic:                              146.1
Date:                Thu, 23 Mar 2023   Prob (F-statistic):                    4.35e-54
Time:                        00:46:06   Log-Likelihood:                         -306.72
No. Observations:                 102   AIC:                                      639.4
Df Residuals:                      89   BIC:                                      673.6
Df Model:                          13                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
CRIM          -0.0773      0.123     -0.630      0.531      -0.321       0.167
ZN             0.1024      0.031      3.260      0.002       0.040       0.165
INDUS         -0.1357      0.175     -0.776      0.440      -0.483       0.212
CHAS           3.5859      2.375      1.510      0.135      -1.133       8.304
NOX            4.1738      9.027      0.462      0.645     -13.762      22.110
RM             4.6533      0.777      5.986      0.000       3.109       6.198
AGE           -0.0139      0.039     -0.358      0.721      -0.091       0.063
DIS           -0.8409      0.477     -1.763      0.081      -1.788       0.107
RAD            0.2129      0.160      1.333      0.186      -0.105       0.530
TAX           -0.0096      0.010     -0.949      0.345      -0.030       0.011
PTRATIO       -0.0775      0.314     -0.247      0.805      -0.701       0.546
B              0.0093      0.007      1.370      0.174      -0.004       0.023
LSTAT         -0.3602      0.151     -2.390      0.019      -0.660      -0.061
==============================================================================
Omnibus:                       64.821   Durbin-Watson:                   1.693
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              394.272
Skew:                           1.987   Prob(JB):                     2.43e-86
Kurtosis:                      11.773   Cond. No.                     1.00e+04
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[3] The condition number is large,  1e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [111]:
# Prediccions en funció de la variable 'LSTAT' (REGRESSIÓ LINEAL - StatsModel)

import numpy as np

# Indiquem el valor de la variable 'LSTAT' per fer la predicció de 'MEDV'
lstat_value = 8 

# Indiquem la posició de la variable 'LSTAT' al dataframe incial
lstat_index = 12

# Calculem les mitjanes per a cada una de les variables
mean_values = x.mean(axis=0)

# Generem una array de '0' amb la mateixa forma del model (analitzat)
array_0 = np.zeros((len(x), 13))

# Indiquem el valor de 'LSTAT' per predir 'MEDV'
array_0[:, lstat_index] = lstat_value

#Indiequem les mitjanes
for i in range(array_0.shape[1]):
    if i != lstat_index:
        array_0[:, i] = mean_values[i]

# Fem la predicció utilitzant el model inicial de Regressio Lineal
predicted_value = model.predict(array_0)

# Generem el valor estiamt de 'MEDV'
print("El valor estimat de 'MEDV' és:", predicted_value[0])

El valor estimat de 'MEDV' és: 24.136640866441716


In [112]:
# Prediccions en funció de la variable 'LSTAT' (LASSO - StatsModel)

import numpy as np

# Indiquem el valor de la variable 'LSTAT' per fer la predicció de 'MEDV'
lstat_value = 8 

# Indiquem la posició de la variable 'LSTAT' al dataframe incial
lstat_index = 12

# Calculem les mitjanes per a cada una de les variables
mean_values = x.mean(axis=0)

# Generem una array de '0' amb la mateixa forma del model (analitzat)
array_0 = np.zeros((len(x), 13))

# Indiquem el valor de 'LSTAT' per predir 'MEDV'
array_0[:, lstat_index] = lstat_value

#Indiequem les mitjanes
for i in range(array_0.shape[1]):
    if i != lstat_index:
        array_0[:, i] = mean_values[i]

# Fem la predicció utilitzant el model inicial de Regressio Lineal
predicted_value = model_lasso_sm.predict(array_0)

# Generem el valor estiamt de 'MEDV'
print("El valor estimat de 'MEDV' és:", predicted_value[0])

El valor estimat de 'MEDV' és: 26.07432039742614


C:\Users\RSIMONNE\Anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but Lasso was fitted with feature names
  warnings.warn(


--------------------------------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------------------------------

### Model 2: Regressió lineal amb SKLearn

In [113]:
from sklearn import linear_model

# Definim les variables independents (feature_names) i la variable depenent (objectiu / target)

variables = Boston_df.drop('MEDV', axis = 1)
objectiu = Boston_df['MEDV']

print(variables)
print(objectiu)


        CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273.0   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273.0   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273.0   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273.0   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273.0   

     PTRATIO       B  LSTAT  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       1

In [114]:
# Ajustem el model

x2 = variables
y2 = objectiu

Regressio_lineal = linear_model.LinearRegression()
model_skl = Regressio_lineal.fit(x2,y2)


In [115]:
# Prediccions:

Prediccions_SKL = Regressio_lineal.predict(x2)

print(Prediccions_SKL[:5]) # mostrem els primers 5 valors obtinguts

[30.00384338 25.02556238 30.56759672 28.60703649 27.94352423]


In [116]:
# Calculem el R2

Regressio_lineal.score(x2,y2)

0.7406426641094095

In [117]:
# Calculem el coeficient de correlació:

Regressio_lineal.coef_

array([-1.08011358e-01,  4.64204584e-02,  2.05586264e-02,  2.68673382e+00,
       -1.77666112e+01,  3.80986521e+00,  6.92224640e-04, -1.47556685e+00,
        3.06049479e-01, -1.23345939e-02, -9.52747232e-01,  9.31168327e-03,
       -5.24758378e-01])

In [118]:
# Calculem la intercepció (intercept): valor de la variable depentent (y) quan el valor de les variables indepentents (x) és igual a '0' 

Regressio_lineal.intercept_

36.45948838509015

In [119]:
 # Calculem el MSE
    
from sklearn.metrics import mean_squared_error
import statsmodels.api as sm

mse = mean_squared_error(y2, Prediccions_SKL)

print("MSE:", mse)

MSE: 21.894831181729202


### Comentaris:

- En aqest cas obtenim un R2 = 0.74. Es una mica mes baix que amb el model anterior, però podriem determinar que te un cert ajust.
- Com en el cas anterior, si ajustessim el model amb variables independents amb coeficient de correlació baixos, probablement el valor R2 disminuiria.
- Observant els ceficients de correlació (coef) podriem intuir quines variables podrien estar més relacionades amb 'MEDV'.
- En aquest cas també podrien ser la columna [5] 'RM', la columa [4] 'NOX' i/o la columna [7] 'LSTAT'

--------------------------------------------------------------------------------------------------------------------------------

In [120]:
# Entrenem el model amb regularització LASSO i Ridge

  # LASSO: especielment utilitzat quan s'espera que, de totes les varibales independents, només siguin rellevants unes poques.   

from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge

model_lasso_skl = Lasso(alpha=1.0)

model_lasso_skl.fit(x2, y2)

Lasso()

In [121]:
# Ridge: especialment utilitzat quan les variables independents estan correlacionades entre sí o quan totes les variables inpendents tenen efecte significatiu sobre la variable objectiu

model_Ridge = Ridge(alpha=1.0)

model_Ridge.fit(x2,y2)

Ridge()

In [122]:
# Prediccions (LASSO):

Prediccions_SKL_Lasso = model_lasso_skl.predict(x2)

print(Prediccions_SKL_Lasso[:5]) # mostrem els primers 5 valors obtinguts

[30.99753918 25.77681736 29.98601449 29.51799813 28.03248999]


In [123]:
# Prediccions (Ridge):

Prediccions_SKL_Ridge = model_Ridge.predict(x2)

print(Prediccions_SKL_Ridge[:5]) # mostrem els primers 5 valors obtinguts

[30.25311604 24.80547336 30.53232402 28.91100981 28.1832052 ]


In [124]:
# Calculem el R2 (LASSO)

model_lasso_skl.score(x2,y2)

0.6825842212709925

In [125]:
# Calculem el R2 (Ridge)

model_Ridge.score(x2,y2)

0.7388703133867616

### Comentaris:

- Al limitar la magnitud dels coeficients del model de regressió pot reduïr l'ajust exceciu del model
- En aquest sentit, pot reduir-se la capacitat del model per ajustar-se perfectament a les dades, el que pot comportar una reducció del R2, tal com observem tant en l'ajust amb LASSO com amb Ridge.
- No obstant, el fet de tenir una R2 menor no implica necessariament que el model sigui pitjor, sino que s'han de tenir en compte altres aspectes.

--------------------------------------------------------------------------------------------------------------------------------

In [126]:
# Prediccions en funció de la variable 'LSTAT' (REGRESSIÓ LINEAL)

import numpy as np

# Indiquem el valor de la variable 'LSTAT' per fer la predicció de 'MEDV'
lstat_value = 8 

# Indiquem la posició de la variable 'LSTAT' al dataframe incial
lstat_index = 12

# Calculem les mitjanes per a cada una de les variables
mean_values = x.mean(axis=0)

# Generem una array de '0' amb la mateixa forma del model (analitzat)
array_0 = np.zeros((len(x), 13))

# Indiquem el valor de 'LSTAT' per predir 'MEDV'
array_0[:, lstat_index] = lstat_value

#Indiequem les mitjanes
for i in range(array_0.shape[1]):
    if i != lstat_index:
        array_0[:, i] = mean_values[i]

# Fem la predicció utilitzant el model inicial de Regressio Lineal
predicted_value = model_skl.predict(array_0)

# Generem el valor estiamt de 'MEDV'
print("El valor estimat de 'MEDV' és:", predicted_value[0])
 

El valor estimat de 'MEDV' és: 24.974540242572857


C:\Users\RSIMONNE\Anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [127]:
# Prediccions en funció de la variable 'LSTAT' (LASSO)

import numpy as np

# Indiquem el valor de la variable 'LSTAT' per fer la predicció de 'MEDV'
lstat_value = 8 

# Indiquem la posició de la variable 'LSTAT' al dataframe incial
lstat_index = 12

# Calculem les mitjanes per a cada una de les variables
mean_values = x.mean(axis=0)

# Generem una array de '0' amb la mateixa forma del model (analitzat)
array_0 = np.zeros((len(x), 13))

# Indiquem el valor de 'LSTAT' per predir 'MEDV'
array_0[:, lstat_index] = lstat_value

# Indiequem les mitjanes
for i in range(array_0.shape[1]):
    if i != lstat_index:
        array_0[:, i] = mean_values[i]

# Fem la predicció utilitzant el model entrenat LASSO
predicted_value = model_lasso.predict(array_0)

# Generem el valor estiamt de 'MEDV'
print("El valor estimat de 'MEDV' és:", predicted_value[0])

El valor estimat de 'MEDV' és: 26.07432039742614


C:\Users\RSIMONNE\Anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but Lasso was fitted with feature names
  warnings.warn(


In [128]:
# Prediccions en funció de la variable 'LSTAT' (RIDGE)

import numpy as np

# Indiquem el valor de la variable 'LSTAT' per fer la predicció de 'MEDV'
lstat_value = 8 

# Indiquem la posició de la variable 'LSTAT' al dataframe incial
lstat_index = 12

# Calculem les mitjanes per a cada una de les variables
mean_values = x.mean(axis=0)

# Generem una array de '0' amb la mateixa forma del model (analitzat)
array_0 = np.zeros((len(x), 13))

# Indiquem el valor de 'LSTAT' per predir 'MEDV'
array_0[:, lstat_index] = lstat_value

# Indiequem les mitjanes
for i in range(array_0.shape[1]):
    if i != lstat_index:
        array_0[:, i] = mean_values[i]

# Fem la predicció utilitzant el model inicial de Ridge
predicted_value = model_Ridge.predict(array_0)

# Generem el valor estiamt de 'MEDV'
print("El valor estimat de 'MEDV' és:", predicted_value[0])

El valor estimat de 'MEDV' és: 25.014486080828803


C:\Users\RSIMONNE\Anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but Ridge was fitted with feature names
  warnings.warn(


------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

### Model 2: Random Forest

In [129]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Característiques del Data Frame

Boston_df.shape  # 506 files - 14 columnes (variables)

(506, 14)

In [130]:
# Separem les varibles independents de la variable que volem predir 'label' (etiqueta)

y = np.array(Boston_df['MEDV'])

# Treiem del DF la variable que volem predir 'label' (y) per establir les variables independents (x)

x = Boston_df.drop('MEDV', axis = 1)

# Guardem els noms de les variables independents (list)

x_list = list(x.columns)

# Convertim el nou DF en array (array amb les variables independents)

x = np.array(x)


In [131]:
# Establim la linia base (baseline), la mitjana que s'haurà de superar amb el nostre model.(de la variable a predir)

y_mean = np.mean(y)

print("Baseline:", y_mean)

baseline_RM = x[:,x_list.index('RM')]

# Calculem els errors de la linia base i la mitjana dels errors:

baseline_errors = abs(baseline_RM - y) 

print("Errors Baseline:", baseline_errors)

print(("Mitjana errors Baseline:"),round(np.mean(baseline_errors), 2))



Baseline: 22.532806324110677
Errors Baseline: [17.425 15.179 27.515 26.402 29.053 22.27  16.888 20.928 10.869 12.896
  8.623 12.891 15.811 14.451 12.104 14.066 17.165 11.51  14.744 12.473
  8.03  13.635  9.058  8.687  9.676  8.301 10.787  8.753 11.905 14.326
  6.987  8.428  7.25   7.399  7.404 12.967 14.159 15.15  18.734 24.205
 27.876 19.83  19.131 18.489 15.131 13.618 14.214 10.57   9.001 13.798
 13.737 14.385 18.489 17.402 13.012 28.151 18.317 24.784 17.155 13.673
 12.959 10.034 15.744 18.238 25.896 17.21  13.613 16.122 11.806 15.015
 17.783 15.739 16.735 17.155 17.827 15.114 13.721 14.66  14.968 14.426
 21.273 17.281 18.498 16.733 17.511 19.97  16.485 16.079 16.593 21.621
 16.183 15.595 16.458 18.789 14.351 21.775 15.237 30.631 35.98  25.784
 20.773 19.719 12.195 13.163 13.933 13.649 13.664 14.273 13.326 13.171
 15.505 16.085 12.887 12.608 12.246 12.372 15.024 13.179 14.528 13.569
 16.13  14.296 14.539 11.444 12.921 15.414 10.087 10.507 11.569  8.663
 12.742 13.274 16.628 12.578  9

In [132]:
from sklearn.ensemble import RandomForestRegressor

# Creem un model Random Forest de 100 arbres de decisió:
rf_Boston = RandomForestRegressor(n_estimators = 100, max_features='sqrt', random_state = 42)

# Ajustem el model 
rf_Boston.fit(x, y)

RandomForestRegressor(max_features='sqrt', random_state=42)

In [133]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

# Predicció del model:

y_pred = rf_Boston.predict(x)

# Calculem l'error quadràtic mitjà del model

mse = mean_squared_error(y, y_pred)

# Calculem la R2 (de la variable que hem de predir)

r2 = r2_score(y, y_pred)

# Resultats de MSE, R2 i la millora sobre la Baseline

print("MSE:", mse)
print("Millora sobre la Baseline:", mean_squared_error(y, np.full_like(y, y_mean)) - mse)
print("R2:", r2)


MSE: 1.5499260138339905
Millora sobre la Baseline: 82.86963014233156
R2: 0.9816402018157164


### Comentaris:

- En aquest cas obtenim un R2 = 0.98, el que ens indica que el model presenta un molt bon ajust.
- El MSE (mean squared error) obtingut és molt inferior als obtinguts als models de regressió lineal (Statsmodel i SKLearn), per tant podríem establir que el model realitzat amb RandomForest és un model més ajustat.
- El model RandomForest és capaç de modelar relacions no lineals entre les variables; aquest podría ser el motiu per el qual untilitzant RandomForest obtenim un model més ajustat.

--------------------------------------------------------------------------------------------------------------------------------

In [134]:
# Entrenem el model utilitzant la tècnica 'Regularized Random Forest' (RRF) amb regularització LASSO 

import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import Lasso
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

importants = rf_Boston.feature_importances_

lasso = Lasso(alpha=0.1)
lasso.fit(x[:,importants>0],y)

y_pred_lasso = lasso.predict(x[:,importants>0])
mse = mean_squared_error(y, y_pred_lasso)

r2 = r2_score(y, y_pred_lasso)

print(mse)
print(r2)

23.04793291321172
0.7269834862602695


----------------------------------------------------------------------------------------------------------------------------

In [135]:
# Fem la partició del conjunt de dades en 'Train' i 'Test'

from sklearn.model_selection import train_test_split

train_x, test_x, train_y, test_y = train_test_split(x, y, test_size = 0.20,random_state = 42)

print('Característiques Train_x:', train_x.shape)
print('Característiques Train_y:', train_y.shape)
print('Característiques Test_x:', test_x.shape)
print('Característiques Test_y:', test_y.shape)

# En 'Train' tenim el 80% dels valors (404 files) i en 'Test' el 20% (102 files)


Característiques Train_x: (404, 13)
Característiques Train_y: (404,)
Característiques Test_x: (102, 13)
Característiques Test_y: (102,)


In [136]:
# Establim la linia base (baseline), la mitjana que s'haurà de superar amb el nostre model.(de la variable a predir)

train_y_mean = np.mean(train_y)
test_y_mean = np.mean(test_y)

print("Linia base Train:", train_y_mean)
print("Linia base Test:", test_y_mean)

baseline_RM = test_x[:,x_list.index('RM')]

# Calculem els errors de la linia base i la mitjana dels errors:

baseline_errors = abs(baseline_RM - test_y) 

round(np.mean(baseline_errors), 2)

   

Linia base Train: 22.796534653465343
Linia base Test: 21.488235294117644


15.34

In [137]:
from sklearn.ensemble import RandomForestRegressor

# Creem un model Random Forest de 1000 arbres de decisió:
rf_Boston = RandomForestRegressor(n_estimators = 1000, random_state = 42)

# Ajustem el model amb les dades d'enrenament ('Train') 
rf_Boston.fit(train_x, train_y)


RandomForestRegressor(n_estimators=1000, random_state=42)

In [138]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

# Prediccions en el models 'Train' i 'Test'

y_pred_train = rf_Boston.predict(train_x)
y_pred_test = rf_Boston.predict(test_x)

# Calculem l'error quadràtic mitjà dels models 'Train' i 'Test' (de la variable que hem de predir 'label')

mse_train = mean_squared_error(train_y, y_pred_train)
mse_test = mean_squared_error(test_y, y_pred_test)

# Calculem la R2 (de la variable que hem de predir 'label')

r2_train = r2_score(train_y, y_pred_train)
r2_test = r2_score(test_y, y_pred_test)


# Resultats de MSE, R2 i la millora sobre la Baseline
print("MSE Train:", mse_train)
print("MSE Test:", mse_test)
print("Millora sobre la Baseline Train:", mean_squared_error(train_y, np.full_like(train_y, train_y_mean)) - mse_train)
print("Millora sobre la Baseline Test:", mean_squared_error(test_y, np.full_like(test_y, test_y_mean)) - mse_test)
print("R2 Train:", r2_train)
print("R2 Test:", r2_test)

MSE Train: 1.8978967982177972
MSE Test: 8.84665565990191
Millora sobre la Baseline Train: 84.97550703473975
Millora sobre la Baseline Test: 64.48712750042104
R2 Train: 0.9781533045272737
R2 Test: 0.8793645264344092


----------------------------------------------------------------------------------------------------------------------------

In [139]:
  # Prediccions en funció de la variable 'LSTAT' (rf_Boston)

import numpy as np

# Indiquem el valor de la variable 'LSTAT' per fer la predicció de 'MEDV'
lstat_value = 8 

# Indiquem la posició de la variable 'LSTAT' al dataframe incial
lstat_index = 12

# Calculem les mitjanes per a cada una de les variables
mean_values = x.mean(axis=0)

# Generem una array de '0' amb la mateixa forma del model (analitzat)
array_0 = np.zeros((len(x), 13))

# Indiquem el valor de 'LSTAT' per predir 'MEDV'
array_0[:, lstat_index] = lstat_value

# Indiequem les mitjanes
for i in range(array_0.shape[1]):
    if i != lstat_index:
        array_0[:, i] = mean_values[i]
        
# Fem la predicció utilitzant el model entrenat LASSO
predicted_value = rf_Boston.predict(array_0)

# Generem el valor estiamt de 'MEDV'
print("El valor estimat de 'MEDV' és:", predicted_value[0])  

El valor estimat de 'MEDV' és: 23.43379999999993


In [140]:
  # Prediccions en funció de la variable 'LSTAT' (LASSO)

import numpy as np

# Indiquem el valor de la variable 'LSTAT' per fer la predicció de 'MEDV'
lstat_value = 8 

# Indiquem la posició de la variable 'LSTAT' al dataframe incial
lstat_index = 12

# Calculem les mitjanes per a cada una de les variables
mean_values = x.mean(axis=0)

# Generem una array de '0' amb la mateixa forma del model (analitzat)
array_0 = np.zeros((len(x), 13))

# Indiquem el valor de 'LSTAT' per predir 'MEDV'
array_0[:, lstat_index] = lstat_value

# Indiequem les mitjanes
for i in range(array_0.shape[1]):
    if i != lstat_index:
        array_0[:, i] = mean_values[i]

# Fem la predicció utilitzant el model entrenat LASSO
predicted_value = lasso.predict(array_0)

# Generem el valor estiamt de 'MEDV'
print("El valor estimat de 'MEDV' és:", predicted_value[0])  

El valor estimat de 'MEDV' és: 25.17932508257821
